README.md
# llamalog
![llamalog Logo](/img/logo_150x150.png) A **l**ightweight **l**ean **a**nd **m**ean **a**synchronous **log**ger for C++.

[![Release](https://img.shields.io/github/v/tag/mbeckh/llamalog?label=Release&style=flat-square)](https://github.com/mbeckh/llamalog/releases/)
[![Tests](https://img.shields.io/github/workflow/status/mbeckh/llamalog/build/master?label=Tests&logo=GitHub&style=flat-square)](https://github.com/mbeckh/llamalog/actions)
[![Coverage](https://img.shields.io/codecov/c/gh/mbeckh/llamalog/master?label=Coverage&logo=codecov&style=flat-square)](https://codecov.io/gh/mbeckh/llamalog)
[![Codacy Grade](https://img.shields.io/codacy/grade/5118cc5d34954c4fa1253f03b3980811?label=Code%20Quality&logo=codacy&style=flat-square)](https://www.codacy.com/manual/mbeckh/llamalog?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=mbeckh/llamalog&amp;utm_campaign=Badge_Grade)
[![License](https://img.shields.io/github/license/mbeckh/llamalog?label=License&style=flat-square)](https://github.com/mbeckh/llamalog/blob/master/LICENSE)

## Feature Highlights
llamalog is basically [NanoLog](https://github.com/Iyengar111/NanoLog) reloaded. It **keeps** the following highlights:

-   All formatting (including any number or object to string conversions) is done in an asynchronous background thread.

-   Log messages of 256 characters or less do not require any heap allocations.

-   Minimalistic header includes to avoid common pattern of (template-heavy) header only library. This helps to keep 
    compilation times of projects low. (BTW: This really is what drew me to NanoLog in the first place).

The **main differences** are:

-   llamalog uses [{fmt}](https://github.com/fmtlib/fmt) for formatting. All formatting is still done asynchronously in 
a background thread.

-   The background thread does not perform near-busy waits in 50 microsecond intervals but instead is notified using
    conditions.

-   Support for formatting custom types (as long as they have a copy constructor). For better efficiency, the logger 
    uses a move constructor if available.

-   Support for wide character strings which are very common in the Windows API. The output is encoded as UTF-8.

-   A time-based rolling file sink.

-   My focus is on Windows desktop development where heap size is not an issue. So I removed the non-guaranteed logger.

-   Shameless use of the Windows API where it provides better performance than the STL.

-   I had to drop the zero-copying for string literals - however it does not seem to hurt performance too badly.

In the course of all the rewriting, I also made some performance improvements, e.g. removing virtual functions because 
there remains only one type of logger, replacing heap-allocated data with stack-based structures and doing some more 
"dirty" stuff.

## Performance
The following benchmarks were compiled using Microsoft Visual Studio 15.9.8 (Microsoft C/C++ Compiler version 
19.16.27027.1) as x64 binaries and executed using Windows 10 on Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz/3.90 GHz. Data
was written to a HDD.

The benchmark compares llamalog to its ancestor [NanoLog](https://github.com/Iyengar111/NanoLog), to
[spdlog](https://github.com/gabime/spdlog) which claims to be very fast and also uses {fmt} and to
[g3log](https://github.com/KjellKod/g3log) which has also been very thoroughly tested for performance.

The following versions were used for the benchmark.

|Project    |Commit                                  |
|---        |---                                     |
|fmt        |287eaab3b2777daa5d0d0cf72d977196ba54efb7|
|g3log      |cb4a94da7ddaec239d6cd95d0dafa0545b3c7c47|
|llamalog   |102713b0d2c9548d20d8c07ee69aa87cc4bbb80d|
|NanoLog    |40a53c36e0336af45f7664abeb939f220f78273e|
|spdlog     |bdfc7d2a5a4ad9cc1cebe1feb7e6fcc703840d71|
|msvc-common|12bcbf336e284294498298b0e7214d9ae28b58aa|

The following tables show the average of three runs for each logger at each setting. Lower numbers are always better.
Percentiles, worst and average duration for a logger call are given in microseconds. The total duration, which also 
includes writing all data to the log file, is stated as seconds.

### 1 Thread - 1,000,000 Messages
llamalog is slightly faster than NanoLog when logging and more than twice as fast when writing which can be accounted to
replacing the C++ stream API with native Windows `WriteFile` calls.

spdlog is slower on average but gives a very consistent behavior with a very low worst case latency. spdlog is also
blazing fast when writing mainly because it uses `fwrite` internally which does very heavy buffering.

|Logger  |50th|75th|90th|99th|99.9th|Worst|Average|   Total|
|:---    |---:|---:|---:|---:|  ---:| ---:|   ---:|    ---:|
|llamalog| 0.1| 0.1| 0.2| 1.4|   3.0|194.2|0.20132| 7.64794|
|NanoLog | 0.2| 0.2| 0.2| 1.4|   3.2|224.9|0.24535|16.20528|
|spdlog  | 0.6| 0.7| 0.8| 3.2|  10.5| 77.8|0.73845| 0.87369|
|g3log   | 2.8| 2.9| 3.7| 7.9|  13.9|415.7|3.08563|12.41233|

### 2 Threads - 500,000 Messages Each (= 1,000,000 Messages in Total)
The performance hit for both llamalog and NanoLog is higher (nearly twice the latency of the single threaded test) than
for spdlog and g3log. However, llamalog and NanoLog deliver a better overall performance.
spdlog is still writing very fast, but the worst case latency does not keep up with the single threaded test. g3log - as
advertised - keeps a very consistent timing and hardly suffers any latency hit at all.

|Logger  |50th|75th|90th|99th|99.9th|Worst|Average|   Total|
|:---    |---:|---:|---:|---:|  ---:| ---:|   ---:|    ---:|
|llamalog| 0.2| 0.2| 1.4| 2.1|   4.3|178.9|0.38859| 7.64205|
|NanoLog | 0.2| 0.3| 1.4| 2.1|   3.8|248.9|0.39784|16.22859|
|spdlog  | 0.7| 0.8| 1.4| 5.2|  16.1|455.5|0.96818| 0.91599|
|g3log   | 2.9| 3.1| 4.4| 7.5|  11.8|558.1|3.17071|12.60405|

### 4 Threads - 250,000 Messages Each (= 1,000,000 Messages in Total)
The trend continues with llamalog and NanoLog again nearly doubling the latency. The same holds true for spdlog and even
g3log now takes a little performance hit but still a lot lower than the other three on average. However, worst case
latency severely suffers for g3log.

|Logger  |50th|75th|90th|99th|99.9th|  Worst|Average|   Total|
|:---    |---:|---:|---:|---:|  ---:|   ---:|   ---:|    ---:|
|llamalog| 0.3| 0.5| 2.0| 3.3|   4.6|  448.9|0.67265| 7.66290|
|NanoLog | 0.3| 0.6| 2.0| 3.2|   4.6|  559.7|0.70082|16.17560|
|spdlog  | 1.1| 1.6| 3.0|12.0|  68.8|  862.8|1.79919| 1.00225|
|g3log   | 4.7| 5.0| 6.0|10.5|  87.1|2,368.7|4.75029|12.55430|

### 10 Threads - 100,000 Messages Each (= 1,000,000 Messages in Total)
At ten threads logging in parallel, timings for all loggers skyrocket. g3log and spdlog still lead the pack when
relative numbers are compared. Both have average latencies of twice the results for 4 threads whereas timings for
llamalog and NanoLog rise by a factor of five. spdlog shines with a worst case latency in a league of its own. This is
also the first test where NanoLog clearly outperforms llamalog.

|Logger  |50th|75th|90th|99th|99.9th|Worst   |Average|   Total|
|:---    |---:|---:|---:|---:|  ---:|    ---:|   ---:|    ---:|
|llamalog| 0.3| 2.7| 3.2| 4.7|   9.0|40,275.1|3.82976| 7.99707|
|NanoLog | 0.3| 2.5| 3.1| 4.6|   6.8|40,967.5|3.34428|16.36179|
|spdlog  | 1.2| 1.9| 3.5|23.6| 221.1|27,640.3|3.55314| 1.04362|
|g3log   | 4.9| 5.5| 7.3|12.1| 162.3|48,077.1|8.66527|12.85125|

### 1 Thread - 1,000,000 Messages - Log Message Longer than 500 Bytes
The following two benchmarks used data for log message longer than 500 bytes to examine the effect when the internal
on-stack buffers are no longer sufficient.

Compared with the single thread test all loggers take a performance hit, yet g3log again showing the smallest relative
change. The total time for NanoLog nearly doubles which shows why the C++ stream API is not well suited for performance
critical applications. spdlog handles this test remarkably well when the 99th and 99.9th percentiles are considered.

|Logger  |50th|75th|90th|99th|99.9th|Worst|Average|   Total|
|:---    |---:|---:|---:|---:|  ---:| ---:|   ---:|    ---:|
|llamalog| 0.5| 0.6| 0.7| 2.8| 100.8|253.9|0.97674| 9.16119|
|NanoLog | 0.6| 0.7| 0.8| 3.2| 103.9|541.3|1.18773|36.64014|
|spdlog  | 1.0| 1.1| 1.2| 3.8|  92.7|447.8|1.27233| 2.66903|
|g3log   | 3.7| 3.8| 4.1|11.3|  17.1|267.6|3.95533|13.51272|

### 4 Threads - 250,000 Messages Each - Log Message Longer than 500 Bytes
Longer messages with 4 threads show the expected results. Average latency for llamalog nearly doubles. This time, even
g3log shows measurable differences. Again, the results for spdlog are very good and the 99th and 99.9th percentiles.

|Logger  |50th|75th|90th|99th|99.9th|  Worst|Average|   Total|
|:---    |---:|---:|---:|---:|  ---:|   ---:|   ---:|    ---:|
|llamalog| 0.7| 1.3| 2.5| 4.6| 146.4|  471.0|1.62763| 9.12183|
|NanoLog | 0.9| 1.2| 2.4| 4.6| 149.4|  647.4|1.78600|37.44443|
|spdlog  | 1.5| 2.0| 3.2| 7.7| 117.6|  555.4|2.19079| 3.17157|
|g3log   | 6.1| 6.4| 7.4|12.8| 110.0|2,123.2|6.09669|13.96604|

### Conclusion
The various loggers are obviously optimized for different use cases. Logging using spdlog and g3log takes longer but
does not change that much with number of parallel threads or message size. On the other hand, both NanoLog and llamalog
generally show the best results up to the 90th percentile.

My aim was not to build the fastest logger but something that is easily usable and fast enough. Based on the data above,
I think llamalog is mature enough to get in the ring. And I can use it for my programming without worrying too much.

## Usage
### Configuration
All configuration is done in code when creating an instance of a `LogWriter`. The global log level SHOULD be set using
on of the preprocessor symbols `LLAMALOG_LEVEL_*`. The value defines the lowest log level for which logging code will be
compiled in the executable. Logging macros for log levels lower than `LLAMALOG_LEVEL_...` are defined as empty macros.
The names are `LLAMALOG_LEVEL_FATAL`, `LLAMALOG_LEVEL_ERROR`, `LLAMALOG_LEVEL_WARN`, `LLAMALOG_LEVEL_INFO`,
`LLAMALOG_LEVEL_DEBUG` and `LLAMALOG_LEVEL_TRACE`  with the default being `LLAMALOG_LEVEL_DEBUG`. Irrespective of the
value, llamalog itself always logs internal messages as `WARN` and `ERROR`. For very special cases, the size of the
internal buffer MAY be set using the preprocessor symbol`LLAMALOG_LOGLINE_SIZE`. Use this when the default size of
256 bytes is too small or large for the arguments in the arguments buffer.

### Basic Example
```cpp
// set global log level to kTrace 
#define LLAMALOG_LEVEL_TRACE

#include <llamalog/llamalog.h>
#include <llamalog/LogWriter.h>
#include <memory>

int main(int argc, char* argv[]) {
    std::unique_ptr<llamalog::RollingFileWriter> writer = 
      std::make_unique<llamalog::RollingFileWriter>(
        // the writer only logs events of this level and above
        llamalog::LogLevel::kTrace,
        // directory for the log files
        "L:\\logs",
        // name of the log file; writer will add date, e.g. llamalog.20190331.log
        "llamalog.log");
    llamalog::Initialize(std::move(writer));
    LOG_INFO("Program {} called with {} arguments.", argv[0], argc);
}
```
This produces the following log line, given the code is stored as `main.cpp` and compiled to `L:\\llamalog.exe` (`9234`
is a sample thread id):
```text
2019-03-31 15:27:12.283 INFO [9234] main.cpp:19 main Program L:\\llamalog.exe called with 1 arguments.
```

## Formatting
The patterns use the standard {fmt} syntax with the following enhancements:
-   Output can be escaped using the syntax for strings in C.
-   Exceptions are supported as formatter arguments.

### Escaping
Per default, any data passed to the logger is written "as is", i.e. including line feeds. Sometimes however, a single
log line is desired, e.g. for parsing log files with simple command line tools like grep. One option is to write a
custom logger which does all the replacement. Another option is to use the included escaping feature.

If escaping for a parameter is set, non-printable characters (i.e. ASCII < 0x20 and the backslash-character are escaped
for arguments passed to the logger. Please note that the logger pattern itself is never modified. To request escaping,
wrap the parameter in `llamalog::escape`, e.g. `llamalog::escape("escape\n")`.

### Exception Formatting
Exceptions can be formatted as first-class arguments, though adding the exception as a logger argument MUST happen
inside the catch clause. If an exception is thrown using llamalog::Throw (and the macro LLAMALOG_THROW respectively)
the result of `what()` is ignored and replaced by the logging message.

The following code shows a valid usage pattern:
```cpp
try {
    THROW(std::invalid_argument("ignored"), "Value {} is invalid", val);
} catch (const std::exception& e) {
    LOG_INFO("An exception has occurred: {}", e);
}
```
It is also possible to log plain C++ exceptions like in the following example. Yet, no extended logging context is
available.
```cpp
try {
    throw std::invalid_argument("somearg");
} catch (const std::exception& e) {
    LOG_INFO("An exception has occurred: {}", e);
}
```

Exceptions are formatted using a default pattern, however a custom pattern MAY be used as in the following example where
only the exception message (i.e. `e.what()`) is logged.
```cpp
try {
    THROW(std::invalid_argument("somearg"));
} catch (const std::exception& e) {
    LOG_INFO("An exception has occurred: {:%w}", e);
}
```
More than one argument specifier MAY be used, e.g. `{:%F:%L}` will print the file name and line number where the
exception happened. The following specifiers are supported. If an exception does not have the information an empty
string is used in the output.

-   `%l` - The log message. If the log message is `nullptr`, `%w` is logged.

-   `%C` - The name of the error category for `std::system_error`, `llamalog::SystemError` and derived classes.

-   `%c` - The error code for `std::system_error`, `llamalog::SystemError` and derived classes.

-   `%m` - The error message for `std::system_error`, `llamalog::SystemError` and derived classes.

-   `%w` - The exception message, i.e. the result of `std::exception` `what()`. If the exception was thrown using
    `llamalog::Throw` (and `LLAMALOG_THROW` respectively) the log message is returned and - in case of
    `std::system_error` and `llamalog::system_error` - includes the error message `%m` (i.e. the output matches 
    `std::system_error` except for the exception argument replaced by the log message).

-   `%T` - The timestamp as `yyyy-MM-dd HH:mm:ss.SSS`.

-   `%t` - The thread id of the thread which has thrown the exception.

-   `%F` - The file name where the exception was thrown.

-   `%L` - The line number where the exception was thrown.

-   `%f` - The function name where the exception was thrown.

Exceptions are often caught using a base class. Therefore it is not known which data is actually available for logging.
Using the pattern syntax above could either lead to ugly output (like in
`An error with code {0:%c} has happened: {0:%m}` when the exception is not a `std::system_error` or overly complex catch
clauses just to log the right data.

llamalog therefore supports conditional patterns which are only printed if any one of the contained specifiers produced
a non-empty output. The pattern from the previous paragraph would produce
`An error with code &nbsp;has happened: somemessage` for a `std::exception` (please also note the double space between
"code" and "has"). The following pattern solves the problem: `An error {0:%[with code %c ]}has happened: {0:%w}` and
will output `An error has happened: somemessage` if the exception is not a `std::system_error`. You may even nest
`%[...]` blocks and you can also use formatter arguments inside these blocks, like in the following example:
```cpp
try {
    // code could throw exceptions with extended logging context and without
} catch (const std::exception& e) {
    LOG_INFO("An exception has occurred: {:%w%[ {0}%F:%L]}", e, "location=");
}
```
This sample logs the exception message and - if extended logging data is available - a space character, the text
"location=" followed by file and line number.


CHANGES.md
# llamalog Changes

## Upcoming Release
-   \[Breaking\] No longer logging all output escaped.
-   \[Feature\] Use automated CI workflow.
-   \[Feature\] Allow custom formatting for null values for strings.
-   \[Fix\] Fixed panic logging triggered too early when logging internal messages from non-logger threads.
-   \[Fix\] Fixed Logger::Flush  not working properly when called right after initialization.
-   \[Fix\] Fixed errors in internal buffer handling.

## v2.0.0 - 2020-06-08
-   \[Breaking\] Renamed LogLevel to Priority.
-   \[Breaking\] Renamed CustomTypes.h to custom_types.h.
-   \[Breaking\] Only escape non-printable characters but leave high ASCII unchanged.
-   \[Breaking\] Changed format string for RECT.
-   \[Feature\] Throw exceptions with formatted messages.
-   \[Feature\] Allow formatting of exceptions, error codes and std::align_val_t.
-   \[Feature\] Null-safe logging of values of pointer variables.
-   \[Feature\] Flush function for logger.
-   \[Feature\] Exception-safe logging by calling SLOG_....
-   \[Feature\] Macro LOG_TRACE_RESULT for logging function return values.
-   \[Feature\] Added LogWriter for sending output to stderr.
-   \[Feature\] Target size of type MAY be provided using macro LLAMALOG_LOGLINE_SIZE for larger of smaller buffer requirements.
-   \[Feature\] Allow setting log level at compile time using LLAMALOG_LEVEL_xxx.
-   \[Optimized\] Reduced foot print of custom types by using a jump table.
-   \[Fix\] Fixed wrong calculation of alignment for log arguments.
-   \[Fix\] Fixed handling of errors during logging.
-   \[Fix\] Compilation error when first argument is a custom argument.
-   \[Fix\] Fixed various other bug found in tests and using clang-tidy.
-   \[Documentation\] Made clear that only parameters are escaped.

## v1.0.0 - 2019-03-15
Initial Release.


src/LogWriter.cpp
/*
Copyright 2019 Michael Beckh

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// @file
/// @copyright Code marked with "from NanoLog" is based on NanoLog (https://github.com/Iyengar111/NanoLog, commit
/// 40a53c36e0336af45f7664abeb939f220f78273e), copyright 2016 Karthik Iyengar and distributed under the MIT License.

/*

Distributed under the MIT License (MIT)

	Copyright (c) 2016 Karthik Iyengar

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in the
Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished
to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

*/

#include "llamalog/LogWriter.h"

#include "llamalog/LogLine.h"
#include "llamalog/Logger.h"
#include "llamalog/winapi_log.h"

#include <fmt/format.h>

#include <windows.h>

#include <algorithm>
#include <atomic>
#include <cassert>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <filesystem>
#include <limits>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

namespace llamalog {

//
// LogWriter
//

LogWriter::LogWriter(const Priority priority) noexcept
	: m_priority(priority) {
	std::atomic_thread_fence(std::memory_order_release);
}

// Derived from `is_logged(LogLevel)` from NanoLog.
bool LogWriter::IsLogged(const Priority priority) const noexcept {
	return priority >= m_priority.load(std::memory_order_relaxed);
}

// Derived from `set_log_level(LogLevel)` from NanoLog.
void LogWriter::SetPriority(const Priority priority) noexcept {
	m_priority.store(priority, std::memory_order_release);
}

// Derived from `to_string(LogLevel)` from NanoLog.
__declspec(noalias) _Ret_z_ char const* LogWriter::FormatPriority(const Priority priority) noexcept {
	switch (priority) {
	case Priority::kTrace:
		return "TRACE";
	case Priority::kDebug:
		return "DEBUG";
	case Priority::kInfo:
		return "INFO";
	case Priority::kWarn:
		return "WARN";
	case Priority::kError:
		return "ERROR";
	case Priority::kFatal:
		return "FATAL";
	default:
		if (static_cast<std::uint8_t>(priority) & 3u) {
			return FormatPriority(static_cast<Priority>(static_cast<std::uint8_t>(priority) & ~3u));
		}
		assert(false);
		return "-";
	}
	__assume(false);
}

namespace {

/// @brief The pattern for formatting a timestamp.
constexpr const char kTimestampPattern[] = "{:04}-{:02}-{:02} {:02}:{:02}:{:02}.{:03}";

/// @brief The size of the output buffer for a formatted timestamp.
constexpr const std::size_t kTimestampOutputBufferSize = 24;

}  // namespace

// Derived from `format_timestamp` from NanoLog.
std::string LogWriter::FormatTimestamp(const FILETIME& timestamp) {
	fmt::basic_memory_buffer<char, kTimestampOutputBufferSize> buffer;
	FormatTimestampTo(buffer, timestamp);
	return fmt::to_string(buffer);
}

template <typename Out>
void LogWriter::FormatTimestampTo(Out& out, const FILETIME& timestamp) {
	SYSTEMTIME st;
	if (!FileTimeToSystemTime(&timestamp, &st)) {
		st = {};  // initialize to all 0
	}
	fmt::format_to(out, kTimestampPattern, st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond, st.wMilliseconds);
}

namespace {

/// @brief Helper function for appending a null-terminated string to a `fmt::basic_memory_buffer`.
/// @tparam Out MUST be of type `fmt::basic_memory_buffer`.
/// @param out The target buffer.
/// @param sz The null-terminated string.
template <typename Out>
inline void Append(Out& out, _In_z_ const char* sz) {
	out.append(sz, sz + std::strlen(sz));
}

/// @brief Default buffer size for log lines.
constexpr std::size_t kDefaultBufferSize = 256;

}  // namespace


//
// StdErrWriter
//

// Derived from `NanoLogLine::stringify(std::ostream&)` from NanoLog.
void StdErrWriter::Log(const LogLine& logLine) {
	fmt::basic_memory_buffer<char, kDefaultBufferSize> buffer;

	FormatTimestampTo(buffer, logLine.GetTimestamp());
	buffer.push_back(' ');
	Append(buffer, FormatPriority(logLine.GetPriority()));
	fmt::format_to(buffer, " [{}] ", logLine.GetThreadId());

	Append(buffer, logLine.GetFile());
	fmt::format_to(buffer, ":{} ", logLine.GetLine());
	Append(buffer, logLine.GetFunction());
	buffer.push_back(' ');

	std::vector<fmt::format_context::format_arg> args;
	logLine.CopyArgumentsTo(args);
	fmt::vformat_to(buffer, fmt::to_string_view(logLine.GetPattern()),
					fmt::basic_format_args<fmt::format_context>(args.data(), static_cast<fmt::format_args::size_type>(args.size())));
	buffer.push_back('\n');
	buffer.push_back('\0');

	fputs(buffer.data(), stderr);
}


//
// DebugWriter
//

// Derived from `NanoLogLine::stringify(std::ostream&)` from NanoLog.
void DebugWriter::Log(const LogLine& logLine) {
	fmt::basic_memory_buffer<char, kDefaultBufferSize> buffer;

	FormatTimestampTo(buffer, logLine.GetTimestamp());
	buffer.push_back(' ');
	Append(buffer, FormatPriority(logLine.GetPriority()));
	fmt::format_to(buffer, " [{}] ", logLine.GetThreadId());

	Append(buffer, logLine.GetFile());
	fmt::format_to(buffer, ":{} ", logLine.GetLine());
	Append(buffer, logLine.GetFunction());
	buffer.push_back(' ');

	std::vector<fmt::format_context::format_arg> args;
	logLine.CopyArgumentsTo(args);
	fmt::vformat_to(buffer, fmt::to_string_view(logLine.GetPattern()),
					fmt::basic_format_args<fmt::format_context>(args.data(), static_cast<fmt::format_args::size_type>(args.size())));
	buffer.push_back('\n');
	buffer.push_back('\0');

	OutputDebugStringA(buffer.data());
}


//
// RollingFileWriter
//

namespace {

/// @brief Data for calculation next roll-over and filename.
struct FrequencyInfo {
	const std::uint64_t breakpoint;  ///< @brief Modulus for calculation the next roll over.
	const char* const pattern;       ///< @brief Pattern for variable part of filename.
};

/// @brief The data for the various `RollingFileWriter::Frequency` values.
constexpr FrequencyInfo kFrequencyInfos[] = {
	// *10[= microseconds] * 1000[= milliseconds] * 1000[= seconds] * 3600[= hours] * 24[= days] * 7[= weeks]
	{10ui64 * 1000 * 1000 * 3600 * 24, "{:0>4}{:0>2}"},              // kMonthly - check every day to make logic easier, code will just re-open the same file every day
	{10ui64 * 1000 * 1000 * 3600 * 24, "{:0>4}{:0>2}{:0>2}"},        // kDaily
	{10ui64 * 1000 * 1000 * 3600, "{:0>4}{:0>2}{:0>2}_{:0>2}"},      // kHourly
	{10ui64 * 1000 * 1000 * 60, "{:0>4}{:0>2}{:0>2}_{:0>2}{:0>2}"},  // kEveryMinute
	{10ui64 * 1000 * 1000, "{:0>4}{:0>2}{:0>2}_{:0>2}{:0>2}{:0>2}"}  // kEverySecond
};
static_assert(sizeof(kFrequencyInfos) / sizeof(kFrequencyInfos[0]) == static_cast<std::uint8_t>(RollingFileWriter::Frequency::kCount));

/// @brief Maximum size of output buffer for the kFrequenceInfos patterns.
constexpr std::size_t kFrequencyOutputBufferSize = 16;

}  // namespace

RollingFileWriter::RollingFileWriter(const Priority priority, std::string directory, std::string fileName, const Frequency frequency, const std::uint32_t maxFiles) noexcept
	: LogWriter(priority)
	, m_directory(std::move(directory))
	, m_fileName(std::move(fileName))
	, m_frequency(frequency)
	, m_maxFiles(maxFiles) {
	// empty
}

RollingFileWriter::~RollingFileWriter() noexcept {
	if (m_hFile != INVALID_HANDLE_VALUE) {  // NOLINT(cppcoreguidelines-pro-type-cstyle-cast): INVALID_HANDLE_VALUE is part of the Windows API.
		if (!CloseHandle(m_hFile)) {
			try {
				LLAMALOG_INTERNAL_WARN("Error closing log: {}", LastError());
			} catch (...) {
				LLAMALOG_PANIC("Error closing log");
			}
		}
	}
}

// @copyright Derived from `NanoLogLine::stringify(std::ostream&)` from NanoLog.
void RollingFileWriter::Log(const LogLine& logLine) {
	const FILETIME timestamp = logLine.GetTimestamp();

	// also try to roll when the file is invalid
	if (CompareFileTime(&m_nextRollAt, &timestamp) != 1 || m_hFile == INVALID_HANDLE_VALUE) {  // NOLINT(cppcoreguidelines-pro-type-cstyle-cast): INVALID_HANDLE_VALUE is part of the Windows API.
		RollFile(logLine);
	}

	fmt::basic_memory_buffer<char, kDefaultBufferSize> buffer;
	FormatTimestampTo(buffer, timestamp);
	buffer.push_back(' ');
	Append(buffer, FormatPriority(logLine.GetPriority()));
	fmt::format_to(buffer, " [{}] ", logLine.GetThreadId());

	Append(buffer, logLine.GetFile());
	fmt::format_to(buffer, ":{} ", logLine.GetLine());
	Append(buffer, logLine.GetFunction());
	buffer.push_back(' ');

	std::vector<fmt::format_context::format_arg> args;
	logLine.CopyArgumentsTo(args);
	fmt::vformat_to(buffer, fmt::to_string_view(logLine.GetPattern()),
					fmt::basic_format_args<fmt::format_context>(args.data(), static_cast<fmt::format_args::size_type>(args.size())));
	buffer.push_back('\n');

	DWORD written;  // NOLINT(cppcoreguidelines-init-variables): Initialized before first use.
	const char* const __restrict data = buffer.data();
	const std::size_t length = buffer.size();
	for (std::size_t position = 0; position < length; position += written) {
		// it will work, however please contact me if you REALLY do log messages whose size does not fit in a DWORD... ;-)
		const DWORD count = static_cast<DWORD>(std::min<std::size_t>(std::numeric_limits<DWORD>::max(), length - position));
		if (!WriteFile(m_hFile, data + position, count, &written, nullptr)) {
			if (m_hFile != INVALID_HANDLE_VALUE) {  // NOLINT(cppcoreguidelines-pro-type-cstyle-cast): INVALID_HANDLE_VALUE is part of the Windows API.
				// no need to log if there is no file
				LLAMALOG_INTERNAL_ERROR("Error writing {} bytes to log: {}", count, LastError());
			}
			// try the next event
			return;
		}
		if (written == length) {
			// spare an addition for the most common case
			break;
		}
	}
}

void RollingFileWriter::RollFile(const LogLine& logLine) {
	const FILETIME timestamp = logLine.GetTimestamp();

	ULARGE_INTEGER nextRollAt = {.LowPart = timestamp.dwLowDateTime, .HighPart = timestamp.dwHighDateTime};

	const FrequencyInfo& frequencyInfo = kFrequencyInfos[static_cast<std::uint8_t>(m_frequency)];
	const std::uint64_t breakpoint = frequencyInfo.breakpoint;
	nextRollAt.QuadPart -= nextRollAt.QuadPart % breakpoint;
	nextRollAt.QuadPart += breakpoint;
	m_nextRollAt.dwLowDateTime = nextRollAt.LowPart;
	m_nextRollAt.dwHighDateTime = nextRollAt.HighPart;

	SYSTEMTIME time;
	if (!FileTimeToSystemTime(&timestamp, &time)) {
		LLAMALOG_INTERNAL_ERROR("Error rolling log: {}", LastError());
		// if we can't get the time, we can't roll the file, try again at next log event
		return;
	}

	std::filesystem::path path(m_directory);
	std::filesystem::path fileName(m_fileName);

	path /= fileName.filename().stem();
	path += '.';
	std::filesystem::path pattern(path);

	fmt::basic_memory_buffer<char, kFrequencyOutputBufferSize> buffer;
	fmt::format_to(buffer, frequencyInfo.pattern, time.wYear, time.wMonth, time.wDay, time.wHour, time.wMinute, time.wSecond);
	path += std::string_view(buffer.data(), buffer.size());
	path += fileName.extension();

	buffer.clear();
	fmt::format_to(buffer, frequencyInfo.pattern, "????", "??", "??", "??", "??", "??", "??");
	pattern += std::string_view(buffer.data(), buffer.size());
	pattern += fileName.extension();

	if (m_hFile != INVALID_HANDLE_VALUE) {  // NOLINT(cppcoreguidelines-pro-type-cstyle-cast): INVALID_HANDLE_VALUE is part of the Windows API.
		if (!CloseHandle(m_hFile)) {
			LLAMALOG_INTERNAL_WARN("Error closing log: {}", LastError());
			// if we can't close the file, leave it open
		}
	}

	// Clean any old log files.
	std::vector<std::wstring> files;
	WIN32_FIND_DATAW findData;
	HANDLE hFindResult = FindFirstFileExW(pattern.c_str(), FindExInfoBasic, &findData, FindExSearchNameMatch, nullptr, FIND_FIRST_EX_LARGE_FETCH);
	if (hFindResult == INVALID_HANDLE_VALUE) {  // NOLINT(cppcoreguidelines-pro-type-cstyle-cast): INVALID_HANDLE_VALUE is part of the Windows API.
		if (const DWORD lastError = GetLastError(); lastError != ERROR_FILE_NOT_FOUND) {
			LLAMALOG_INTERNAL_WARN("Error deleting log: {}", error_code{lastError});
		}
	} else {
		do {
			files.emplace_back(findData.cFileName);
		} while (FindNextFileW(hFindResult, &findData));  // NOLINT(readability-implicit-bool-conversion): Compare BOOL in a condition.
		const DWORD error = GetLastError();
		if (error != ERROR_NO_MORE_FILES) {
			LLAMALOG_INTERNAL_WARN("Error deleting log: {}", error_code{error});
			// do not try to delete anything on find errors
		} else {
			std::sort(files.begin(), files.end());
			const std::uint32_t len = static_cast<std::uint32_t>(files.size());
			for (std::uint32_t i = 0; i < len - std::min(len, m_maxFiles); ++i) {
				std::filesystem::path file = std::filesystem::path(m_directory) / files[i];
				if (!DeleteFileW(file.c_str())) {
					LLAMALOG_INTERNAL_WARN("Error deleting log '{}': {}", files[i], LastError());
					// delete failed, but continue
				}
			}
		}
		if (!FindClose(hFindResult)) {
			LLAMALOG_INTERNAL_WARN("Error deleting log: {}", LastError());
			// just leave the handle open
		}
	}

	m_hFile = CreateFileW(path.c_str(), FILE_APPEND_DATA, FILE_SHARE_READ, nullptr, OPEN_ALWAYS, FILE_ATTRIBUTE_NORMAL | FILE_FLAG_SEQUENTIAL_SCAN, nullptr);
	if (m_hFile == INVALID_HANDLE_VALUE) {  // NOLINT(cppcoreguidelines-pro-type-cstyle-cast): INVALID_HANDLE_VALUE is part of the Windows API.
		LLAMALOG_INTERNAL_ERROR("Error creating log: {}", LastError());
		// no file created, try again for next message
	}

	// using CreateFile with FILE_APPEND_DATA does not require a SetFilePointer to the end
}

}  // namespace llamalog
